rules/rules_test.go

```diff
@@ -667,6 +673,151 @@ func TestExpressions(t *testing.T) {
+			output: []string{
+				`{instance="0", job="api-server"} => 3 @[%v]`,
+				`{instance="0", job="app-server"} => 1.4 @[%v]`,
+				`{instance="1", job="api-server"} => 2 @[%v]`,
+				`{instance="1", job="app-server"} => 1.3333333333333333 @[%v]`,
+			},
+		},
+		{
+			// Include labels must guarantee uniquely identifiable time series.
+			expr:   `http_requests{group="production"} / on(instance) group_left(group) cpu_count{type="smp"}`,
+			output: []string{}, // Empty result returned on error (see TODOs).
+		},
+		{
+			// Many-to-many matching is not allowed.
+			expr:   `http_requests{group="production"} / on(instance) group_left(job,type) cpu_count`,
+			output: []string{}, // Empty result returned on error (see TODOs).
+		},
+		{
+			// Include labels must only be available on one side.
+			expr:   `http_requests{group="production"} / on(instance) group_left(job) http_requests{instance="0",group="canary"}`,
+			output: []string{}, // Empty result returned on error (see TODOs).
+		},
+		{
+			expr: `http_requests{group="production"} / on(instance) group_left(job) cpu_count{type="smp"}`,
+			output: []string{
+				`{instance="1", job="api-server"} => 1 @[%v]`,
+				`{instance="0", job="app-server"} => 5 @[%v]`,
+				`{instance="1", job="app-server"} => 3 @[%v]`,
+				`{instance="0", job="api-server"} => 1 @[%v]`,
+			},
+		},
+		{
+			// Ensure sidedness of grouping preserves operand sides.
+			expr: `cpu_count{type="smp"} / on(instance) group_right(job) http_requests{group="production"}`,
+			output: []string{
+				`{instance="1", job="app-server"} => 0.3333333333333333 @[%v]`,
+				`{instance="0", job="app-server"} => 0.2 @[%v]`,
+				`{instance="1", job="api-server"} => 1 @[%v]`,
+				`{instance="0", job="api-server"} => 1 @[%v]`,
+			},
+		},
+		{
+			// Include labels from both sides.
+			expr: `http_requests{group="production"} / on(instance) group_left(job,type) cpu_count{type="smp"}	`,
+			output: []string{
+				`{instance="1", job="api-server", type="smp"} => 1 @[%v]`,
+				`{instance="0", job="app-server", type="smp"} => 5 @[%v]`,
+				`{instance="1", job="app-server", type="smp"} => 3 @[%v]`,
+				`{instance="0", job="api-server", type="smp"} => 1 @[%v]`,
+			},
+		},
+		{
+			// Missing label list for grouping mod.
+			expr:       `http_requests{group="production"} / on(instance) group_left cpu_count{type="smp"}`,
+			shouldFail: true,
+		},
+		{
+			// No group mod allowed for logical operations.
+			expr:       `http_requests{group="production"} or on(instance) group_left(type) cpu_count{type="smp"}`,
+			shouldFail: true,
+		},
+		{
+			// No group mod allowed for logical operations.
+			expr:       `http_requests{group="production"} and on(instance) group_left(type) cpu_count{type="smp"}`,
+			shouldFail: true,
+		},
+		{
+			// No duplicate use of label.
+			expr:       `http_requests{group="production"} + on(instance) group_left(job,instance) cpu_count{type="smp"}`,
+			shouldFail: true,
```

Comment:
For completeness can we verify that on clauses with scalars are an error, and 1-2 tests of the comparison operators too?